snapcraft/internal/lifecycle.py

```diff
@@ -203,12 +204,36 @@ def _create_meta(self, step, part_names):
                                        self.project_options.snap_dir,
                                        self.project_options.parts_dir)
 
-    def _handle_dirty(self, part, step):
+    def _handle_dirty(self, part, step, dirty_report):
         if step not in _STEPS_TO_AUTOMATICALLY_CLEAN_IF_DIRTY:
-            raise RuntimeError(
-                'The {0!r} step of {1!r} is out of date. Please clean that '
-                "part's {0!r} step in order to rebuild".format(
-                    step, part.name))
+            message_components = [
+                'The {!r} step of {!r} is out of date:\n\n'.format(
+                    step, part.name)]
+
+            if dirty_report.dirty_properties:
+                humanized_properties = formatting_utils.humanize_list(
+                    dirty_report.dirty_properties, 'and')
+                pluralized_connection = formatting_utils.pluralize(
+                    dirty_report.dirty_properties, 'property appears',
+                    'properties appear')
+                message_components.append(
+                    'The {} part {} to have changed.\n'.format(
+                        humanized_properties, pluralized_connection))
+
+            if dirty_report.dirty_project_options:
+                humanized_options = formatting_utils.humanize_list(
+                    dirty_report.dirty_project_options, 'and')
+                pluralized_connection = formatting_utils.pluralize(
+                    dirty_report.dirty_project_options, 'option appears',
+                    'options appear')
+                message_components.append(
+                    'The {} project {} to have changed.\n'.format(
+                        humanized_options, pluralized_connection))
```

Comment:
just a note that this is not nice for translations. It's ok for now because we are not doing translations, but hopefully the tool we choose for our messages will make this nicer. https://docs.python.org/3/library/gettext.html#gettext.ngettext